pystiche_papers/sanakoyeu_et_al_2018/_modules.py

```diff
@@ -184,3 +192,202 @@ def residual_block(channels: int) -> ResidualBlock:
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_4": ConvBlock(
+                in_channels=512,
+                out_channels=512,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_5": ConvBlock(
+                512,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+            "scale_6": ConvBlock(
+                1024,
+                1024,
+                kernel_size=5,
+                stride=2,
+                padding="same",
+                act="lrelu",
+                inplace=inplace,
+            ),
+        }
+    )
+    return cast(Dict[str, nn.Sequential], modules)
+
+
+class DiscriminatorEncoder(MultiLayerEncoder):
+    r"""Encoder part of the Discriminator from :cite:`SKL+2018`.
+
+    Args:
+        in_channels: Number of channels in the input. Defaults to ``3``.
+    """
+
+    def __init__(self, in_channels: int = 3) -> None:
+        super().__init__(
+            self._collect_modules(
+                discriminator_encoder_modules(in_channels=in_channels)
+            )
+        )
+
+    def _collect_modules(
+        self, wrapped_modules: Dict[str, nn.Sequential]
+    ) -> List[Tuple[str, nn.Module]]:
+        modules = []
+        block = 0
+        for sequential in wrapped_modules.values():
+            for module in sequential._modules.values():
+                if isinstance(module, nn.Conv2d):
+                    name = f"conv{block}"
+                elif isinstance(module, nn.InstanceNorm2d):
+                    name = f"inst_n{block}"
+                else:  # isinstance(module, nn.LeakyReLU):
+                    name = f"lrelu{block}"
+                    # each LeakyReLU layer marks the end of the current block
+                    block += 1
+
+                modules.append((name, module))
+        return modules
+
+
+def prediction_module(
+    in_channels: int, kernel_size: Union[Tuple[int, int], int], padding: str = "same"
+) -> nn.Module:
+    r"""Prediction module from :cite:`SKL+2018`.
+
+    This block comprises a convolutional, which is used as an auxiliary classifier to
+    capture image details on different scales of the :class:`DiscriminatorEncoder`.
+
+    Args:
+        in_channels: Number of channels in the input.
+        kernel_size: Size of the convolving kernel.
+        padding: Padding of the input. It can be either be``"valid"`` for no padding or
+            ``"same"`` to keep the size. Defaults to ``"same"``.
+
+    """
+    return conv(
+        in_channels=in_channels,
+        out_channels=1,
+        kernel_size=kernel_size,
+        stride=1,
+        padding=padding,
+    )
+
+
+def get_prediction_modules() -> Dict[str, nn.Module]:
```

Comment:
What would you use this for?

Reply:
I use this later to set the default `prediction_modules` from the paper in the `discriminator_operator`. I think this will be clearer if the file _loss.py is integrated.

Reply:
In that case, please remove this for now and only implement this when its needed.